### PR TITLE
feat: migrate only changed and archive

### PR DESCRIPTION
**Pull request by Davidmattei**
|Q              |A  |
|---------------|---|
|Bug fix?       |n|
|New feature?   |y|	
|BC breaks?     |n|	
|Deprecations?  |n|	
|Fixed tickets  |n|	

# Summary
- Add option **update-referrers** in ems:environment:create
- Add option **changed** in ems:contenttype:migrate for only migrating changed revisions
- Add option **archive** in ems:contenttype:migrate for archiving untouched revisions after migration
- Add new command **ems:revision:archive** for archive all revisions for a contentType
- Created a new Revisions Aggregate object, build by repository search method.
- Added **getHash()** shortcut method on Revision
- Removed snapshot parameter from EnvironmentService. Environments becomes a snapshot after running align with the option *snapshot* passed.
- Add giveByName on ContentTypeService

### CODE REVIEW

**Review by theus77 — APPROVED** [review]

File: src/Entity/Revision.php [hunk 1]
```diff
@@ -50,6 +50,13 @@ class Revision
      */
     private $autoSaveAt;
 
+    /**
+     * @var bool
+     *
+     * @ORM\Column(name="archived", type="boolean")
```
Comment (theus77) [inline src/Entity/Revision.php:56]: Specify a default value for doctrine?

File: src/Repository/RevisionRepository.php [hunk 2]
```diff
@@ -48,6 +49,32 @@ public function findByContentType(ContentType $contentType, $orderBy = null, $li
...
+        if (isset($search['lockBy'])) {
+            $qb->andWhere($qb->expr()->eq('r.lockBy', ':lock_by'))->setParameter('lock_by', $search['lockBy']);
+        }
+        if (isset($search['archived'])) {
+            $qb->andWhere($qb->expr()->eq('r.archived', ':archived'))->setParameter('archived', $search['archived']);
+        }
+
+        return new Revisions($qb);
+    }
+
```
Comment (theus77) [inline src/Repository/RevisionRepository.php:77]: I don't get it. why not a spefic searchFunciton searchRevision(?ContentType $contentTypee, ...). This is overriding the doctrine method and can introduce regression I'm afraid.

Comment (Davidmattei) [reply src/Repository/RevisionRepository.php:77]: There is no doctrine method "search", it is not overriding anything.

File: src/Resources/DoctrineMigrations/pdo_sqlite/Version20210526190542.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        , auto_save CLOB DEFAULT NULL COLLATE BINARY --(DC2Type:json_array)
+        , circles CLOB DEFAULT NULL COLLATE BINARY --(DC2Type:simple_array)
+        , labelField CLOB DEFAULT NULL COLLATE BINARY, sha1 VARCHAR(255) DEFAULT NULL COLLATE BINARY, version_uuid CHAR(36) DEFAULT NULL COLLATE BINARY --(DC2Type:uuid)
+        , version_tag VARCHAR(255) DEFAULT NULL COLLATE BINARY, ouuid VARCHAR(255) DEFAULT NULL COLLATE utf8_bin, archived BOOLEAN DEFAULT \'0\' NOT NULL, archived_by VARCHAR(255) DEFAULT NULL, CONSTRAINT FK_6D6315CC1A445520 FOREIGN KEY (content_type_id) REFERENCES content_type (id) NOT DEFERRABLE INITIALLY IMMEDIATE)');
```
Comment (theus77) [inline src/Resources/DoctrineMigrations/pdo_sqlite/Version20210526190542.php:25]:
```suggestion
        , version_tag VARCHAR(255) DEFAULT NULL COLLATE BINARY, ouuid VARCHAR(255) DEFAULT NULL COLLATE BINARY, archived BOOLEAN DEFAULT \'0\' NOT NULL, archived_by VARCHAR(255) DEFAULT NULL, CONSTRAINT FK_6D6315CC1A445520 FOREIGN KEY (content_type_id) REFERENCES content_type (id) NOT DEFERRABLE INITIALLY IMMEDIATE)');
```

**Pull request merged**